# The One Factor Model

The One Factor model is a natural extension of univariate genome-wide association study (GWAS).

In standard GWAS analyses, a single phenotype is repeatedly regressed on each single nucleotide polymorphism (SNP) in the users genomic data file (as well as a fixed set of covariates). The One Factor Model in GW-SEM  allows users to regress a latent factor on each SNP (as well as a fixed set of covariates). This allows users to explicitly construct a measurement model (i.e., factor analysis) within the GWAS context, and go beyond thinking of phenotypes as directly measured variables or case-control status, opening up a whole array of innovative applications. A general schematic depiction of the one factor model is presented in the figure below, but users should keep in mind that the number of items in the analysis and the specific parameterization of the model can be adjusted to suit the specific research question. 

In the Figure below, the latent factor (F<sub>1</sub>) is the underlying (unmeasured) cause the observed items (x<sub>k</sub>). The association between the latent factor and the observed indicators are estimated by the factor loadings (&lambda;<sub>k</sub>). The residual variances (&delta;<sub>k</sub>) indicate the remaining variance in (x<sub>k</sub>) that is not shared with the latent factor. The regression of the latent factor on the SNP (for all SNPs in the analysis) is depicted by (&beta;<sub>k</sub>). For a one factor model to be identified, you will need at least three items, but more items will improve the accuracy of the latent variable. While there are statistical tricks that can be used to identify latent variables with two indicators, GW-SEM does not offer ready-made models that incorporate these tricks. 

***

![Schematic Depiction of the One-Factor GWAS Model](https://github.com/bradverhulst/gwsem/blob/master/exampleFigs/singleFactor.jpg)


***
## Example
To demonstrate how to use these function, we use simulated data for a one factor GWAS model with 6,000 individuals consisting of 2,000 SNPs, three correlated items and 6 covariates (that are a proxy for age, sex, ancestry principle components or other confounds). The simulated data are loosely based upon the data used in the demonstration of Pritikin et al (under review). In the paper, we examined the association a latent substance use frequency variable that used tobacco, cannabis, and alcohol as indicators.  The latent variable was regressed on each SNP with a minor allele frequency larger than 1%.  As we are not allowed to post the raw data online, we simulated similar data that for this demonstration. The phenotypic data data and scripts that we used in this demonstration can be found [here](https://github.com/bradverhulst/gwsem/tree/master/gwsemOneFacExample). The GWAS data can be found [here](https://github.com/bradverhulst/gwsem/tree/master//Example%20GWAS%20Data).

## Running the One-Factor GWAS

The first step is opening R and loading the GW-SEM package into the R computing environment (which will also load all of the dependencies, such as OpenMx), which can be achieved by using the command below:

```ruby
require(gwsem)
```

The next step is to load the phenotypic data. This will include any items that define the latent variable or necessary covariates that you will include in the analysis, such as age, sex, or ancestry principle components. Once the phenotypic data has been loaded, it is advisable to make sure the data has been loaded properly. This can be done in a number of ways. For example, you can use the `head` command to inspect the first few lines of data, look at the means and variances, construct tables of ordinal variables, etc.  If your data didn't load properly, it will likely cause problems for the analysis and for the interpretation of the results.

```ruby
phenoData <- read.table("phenoData.txt", header=TRUE)
head(phenoData)
```

Once the data is loaded into R, you can recode the data, transform it, and tell R if we have ordinal or binary indicators using mxFactor(). The data for this example are simulated and continuous, and therefore, we will not be doing anything now, but if you would like to `cut` the indicators up into binary or ordinal variables, this would be when you would do it.


After you are satisfied the data is in the appropriate shape we can build a one factor GWAS model with the following command: 

```ruby
                                                                                     # You must tell GW-SEM:
addFac <- buildOneFac(phenoData = phenoData,                                         # what the data object is (which you read in above)
                     itemNames = c("tobacco", "cannabis", "alcohol"),                # what the items of the latent factor are
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),                    # what covariates that you want to include in the analysis
                     fitfun = "WLS")                                                 # and the fit function that you would like to use (WLS is much faster than ML)

```

You can take this object (addFac) which is technically an OpenMx model, and simply fit it using mxRun() 

```ruby
addFacFit <- mxRun(addFac)
summary(addFacFit)
```

This is strongly advised, as it is a great time to test whether the model is being specified the way you want it to be and that you are not getting unrealistic estimates.

Provided that the model looks reasonable, you can plug the model that you have built into the GWAS function using the command below:

```ruby
GWAS(model = addFac,                                                                 # what model object you would like to fit
	snpData = 'example.pgen',                                                        # that path to the snpData file.
	out="latFac.log",                                                                # the file that you would like to save the full results into
	SNP=1:200)                                                                       # the index of the snps (how many) you would like to fit
```

**Note about the snpData argument:** The path to your snpData will likely include switching to a different directory (as you will likely do your analysis in a different folder than your SNP data). All you need to do is point to the data using relative paths. Further, it is able to take plink bed/bim/fam or pgen/psam/pvar data or bgen data (Oxford format)

**Note about the SNP argument:** This can be used to run a limited number of SNP (i.e. not the whole snp file). This is particularly useful if you would like to chop up a chr into several parts without cutting you actual genotype data into separate files.

While the GWAS function will take a while and frequently be executed on a computing cluster, it is very useful to run a few SNPs (say 10 or 50) in an interactive session to make sure that: all of your relative file paths to the genotypes are correct, the model is taking a reasonable amount of time (i.e., 1-2 seconds/snp), the SNPs are giving sensible estimates, etc., as the results from a few SNPs can often tell you if there is a problem and that you are running a nonsensical model. 

While most computing clusters have unique queue submission systems (and I would strongly advise you to run GWAS on a computing cluster), the R script that you provide will typically be similar to the one below:

```ruby
require(gwsem)

phenoData <- read.table("phenoData.txt", header=TRUE)
head(phenoData)

addFac <- buildOneFac(phenoData = phenoData,                                         # what the data object is (which you read in above)
                     itemNames = c("tobacco", "cannabis", "alcohol"),                # what the items of the latent factor are
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),                    # what covariates that you want to include in the analysis
                     fitfun = "WLS")                                                 # and the fit function that you would like to use (WLS is much faster than ML)

GWAS(model = addFac,                                                                 # what model object you would like to fit
	snpData = 'example.pgen',                                                    # that path to the snpData file.
	out="latFac.log")                                                            # the file that you would like to save the full results 
```

The next step is to read the results into R. The output from the GWAS function contains all of the estimates and standard errors for all parameters in the model (for each SNP), as well as other general model fitting information.  While you are unlikely to want to read all of this output into R, it is useful to do this on a small test GWAS analysis to ensure that you are getting reasonable results and to familiarize yourself with what information is available, in case you need to explore a specific parameter or model in more detail. This can be done using the command below:  

```ruby
FullResult <- read.delim("latFac.log")
```

More likely, you are going to want to read the results into R for SNP regression coefficients. This function takes takes two arguments: the path to the data and the column in the results file for the parameter that you want to examine.

```ruby
succinct <- loadResults(path = "latFac.log", focus =  "snp_to_F")
```

We can then construct a Manhattan plot using the GW-SEM plot() function. 

```ruby
plot(succinct)
```

The loadResults() function formats the summary statistics from the analyses so that the data object can be seamlessly used by qqman, LDhub, and other post GWAS processing software.



***

## Suggestions for Conducting the Analysis

Before running any sort of analysis in GW-SEM, users should have an understanding of the phenotypic relationships between the traits that you are interested in examining. Specifically, for the one-factor model, the items that load onto the latent variable should be established prior to running the GWAS, or the user risks conducting an analysis that does not test their intended hypothesis.  For example, if the structure of the latent factor differs from the model written in GW-SEM (model misspecification), any conclusions that are drawn from the analysis are questionable.  Alternatively, if there are items that do not load onto the latent factor, this may dilute the GWAS signal.  The best, and most interpretable, results will come from a clearly defined, well specified, phenotypic model. 

***

### Using the Functions and Understanding the Arguments

The GW-SEM workflow is split up into two _primary_ functions.  The first function specifies the model and the second function runs the GWAS analysis.


### buildOneFac

The **buildOneFac** function helps users easily specify a one-factor model that can be passed to the GWAS function. In essence, users declare which items load on the latent variable and which variable should be treated as covariates, and GW-SEM will build a one-factor model with all of the components needed to conduct a GWAS. The syntax of the **buildOneFac** model is presented below:

```ruby
buildOneFac(phenoData, itemNames, covariates=NULL, ..., fitfun = c("WLS","ML"), minMAF=0.01, , gxe = NULL, exogenous = NA)
```
**phenoData**: the file pathway for the phenotypic data (e.g. "myData.txt" or "phenotype/myData.txt"). This data file can include more variables than those included in the analysis, but GW-SEM will only use the items/covariates that are specified.  (The dangers of very large dataset is that they can take a long time to load and can take up space in the R environment.  This should not affect processing speed for the GWAS analysis, but can create headaches for pre-processing).

**itemNames**: a character list of the names of the items that load onto the latent variable.  These names must match variable names in the **phenoData** file.

**covariates**: a character list of covariates that the latent variable will be regressed upon.  The default vaule is **NULL**, but this is a silly value as typically analysts will include e.g. age, sex, and ancestry principle components in the analysis.

**fitfun**:  The **fitfun** argument specifies which fit function should be used in evaluating the GWAS model. Users may choose between the relatively rapid "WLS", or the slower but asymptotically optimal "ML". In many cases the the differences between the fit functions is trivial and the faster "WLS" option should be used, but in some situations the differences can be quite meaningful (such as when data are Missing at Random - MAR).

**minMAF**: The minimum valid minor allele frequency (MAF).  Large differences between the variances of two variables in the same model can cause optimization failures that invalidate the model.  Further, very small minor allele frequencies are more affected by outliers or influential observations.  Accordingly, users can specify the minimum allowable MAF. The default value is MAF > .01.  Users may also wish to filter out small MAF alleles from their genotype files in other software programs, such as PLINK.

**gxe:** The observed variable name that will be used to moderate (interact with) the effect of the SNP on the phenotypes.  For example, you may want to moderate the SNP regression by sex.  In this situation, you would specify **_gxe = "sex"_** and add _**"snp_sex"**_ to the list of covariates.

**exogenous:** This argument specifies how you would like to integrate the covariates into the analysis. If **exogenous = T**, each items will be directly regressed on each covariate. If **exogenous = F**, the latent factor(s) will be directly regressed on each covariate. Setting **exogenous = T** does not assume that the items are related to the covariates proportional to their factor loadings (which is probably preferable in most cases).

### Running the GWAS

The **GWAS** function takes the model that was created by  **buildOneFac** (or the other prebuilt or user-specified GW-SEM functions) and conducts a GWAS.  To run the GWAS, use the **GWAS** function. The syntax of the **GWAS** model is presented below:

```ruby
GWAS(model, snpData, out="out.log", SNP=NULL, startFrom = 1L)
```
**model**: an OpenMx model, specified using RAM or LISREL notation.  The model argument is designed to take the output from e.g. **buildOneFac** (or the other prebuilt GW-SEM functions), but advanced users can specify their own arbitrary OpenMx Model or use Onyx to draw their path diagrams.

**snpData**: a pathway to a file containing GWAS data.  The data can be in a variety of forms, such as standard PLINK format (bed/bim/fam), PLINK2 format (pgen/pvar/psam), Oxford format (bgen/sample), or CSV format (csv format in much slower due to the lack of compression for non-binary files).

**SNP** : a numerical range that specifies the number of SNPs to be evaluated from the **snpData** file. This argument can be used to evaluate a subset of snps for model testing. e.g. 1:10 will run the first 10 snps to make sure that the model is functioning the way the users intends, that the files exist pathways are correct. This option is also very useful to specify a range of snps to be evaluated that is smaller than the complete file.  For example, users may wish to run several discrete batches of analyses for chromosome 1, by running 1:10000, 100001:200000, etc.  This prevents users from constructing numerous snap files for each chromosome. The default value of the **SNP** argument is NULL, which will run all snps in the file.

**out** : a file name or pathway where the output from the analysis will be saved. The default pathway is "out.log", which will save the file in the working directory. Users should take caution when specifying the output file name so that the output from different analyses/chromosomes do not overwrite existing files.

**startFrom** : a numerical value indicating which SNP is the first SNP to be analyzed.  The function will then run every SNP from the specified SNP to the end of the GWAS data file. This is very useful if the analysis stops for some reason (i.e. the cluster is restarted for maintenance) and you can start from the last SNP that you analyzed. Note, you will want to label the output file (specified in **out**) with a new file name so that you don't overwrite the existing results.