# The Two Factor Model
***

The Two Factor model is an extension of the [One Factor Model](https://github.com/bradverhulst/gwsem/wiki/One-Factor-Model). The obvious difference between the one and two factor models are the additional regression coefficient and the correlations between the two latent variables (and the additional factor of course).

Unlike the One Factor GWAS model, the Two Factor GWAS model does not have an analogue in the current GWAS literature but nevertheless plays an integral role the many genetic theories of comorbidity and pleiotropy. For example, multivariate twin models routinely detect genetic variation across a range of phenotypes, and interpret these findings in terms of genetic variants that are shared between the phenotypes.  LD score regression analyses take this one step further, testing for genetic overlap between phenotypes based on summary statistics from univariate GWAS.  The two factor GWAS model takes this theoretical hypothesis further than existing methods by allowing users to explicitly test for pleiotropy in a single model. This framework allows users to think about phenotypes from an explicitly multivariate perspective that cannot be fit with alternative GWAS software. A general schematic depiction of the two factor model is presented in the figure below, but users should keep in mind that the number of items in the analysis and the specific parameterization of the model can be adjusted to suit the a specific research question. 

In the Figure below , the latent factor (F<sub>1</sub>) is the underlying (unmeasured) cause the observed items (x<sub>k</sub>). The association between the latent factor and the observed indicators, (x<sub>1</sub>, x<sub>2</sub> \& x<sub>3</sub>), are estimated by the factor loadings (&lambda;<sub>1</sub>, &lambda;<sub>2</sub>, \& &lambda;<sub>3</sub>, respectively). The residual variances (&delta;<sub>1</sub>, &delta;<sub>2</sub>, \& &delta;<sub>3</sub>) indicate the remaining variance in (x<sub>k</sub>) that is not shared with the latent factor. The regression of the latent factor on the SNP (for all SNPs in the analysis) is depicted by (&beta;<sub>F1</sub>). The corresponding factor loadings for the second latent factor  (F<sub>1</sub>) and the y<sub>k</sub> variables can be interpreted similarly. The new parameter that is added to the two factor model that was absent from the one factor model is the covariance between the latent variables (&psi;). This parameter can be interpreted in conjunction with the &beta;<sub>F1</sub> and &beta;<sub>F2</sub> parameters to assess the impact of the genomic associations on comorbidity.

For the two factor model to be identified, you will need at least three items for each factor (as was the case with the one factor model), but more items will improve the measurement of the latent variables. 

![A schematic depiction of the Two-Factor GWAS Model](https://github.com/bradverhulst/gwsem/blob/master/exampleFigs/twoFactor.jpg)
***

## Example

To demonstrate how to conduct a Two Factor GWAS, we use the same simulated GWAS data for a one factor GWAS model with 6,000 individuals consisting of 2,000 SNPs, and simulated novel phenotypic data with six items (A<sub>1</sub>, A<sub>2</sub> \& A<sub>3</sub> for F<sub>1</sub>, and B<sub>1</sub>, B<sub>2</sub> \& B<sub>3</sub> for F<sub>2</sub>) as well as 6 covariates (that are a proxy for age, sex, ancestry principle components or other confounds). The latent variables F<sub>1</sub> and F<sub>2</sub> were simultaneously regressed on each SNP. The phenotypic data data and scripts that we used in this demonstration can be found [here](https://github.com/bradverhulst/gwsem/tree/master/gwsemTwoFacExample). The GWAS data can be found [here](https://github.com/bradverhulst/gwsem/tree/master//Example%20GWAS%20Data).


## Running the Two-Factor GWAS

The first step is opening R and loading the GW-SEM package into the R computing environment (which will also load all of the dependencies, such as OpenMx), which can be achieved by using the command below:

```ruby
require(gwsem)
```

The next step is to load the phenotypic data. This will include any items that define the latent variables or necessary covariates that you will include in the analysis, such as age, sex, or ancestry principle components. Once the phenotypic data has been loaded, it is advisable to make sure the data has been loaded properly by, e.g., inspecting the first few lines of data, looking at the means and variances, etc.  This can be done with the code below:

```ruby
TwoDat <- read.table("phenoTwoData.txt", header=TRUE)
head(TwoDat)
```

Once the data is loaded into R, you can recode the data, transform it, and tell R if we have ordinal or binary indicators using mxFactor(). The data for this example are simulated and continuous, and therefore, we will not be doing anything now, but if you would like to chop the indicators up into binary or ordinal variables, this would be when you would do it.


After you are satisfied the data is in the appropriate shape we can build a two factor GWAS model with the following command: 

```ruby

twoFac <- buildTwoFac(phenoData = TwoDat,                                         # what the data object is (which you read in above)
                     F1itemNames = c("A1", "A2", "A3"),                # what the items of the latent factor are
                     F2itemNames = c("B1", "B2", "B3"),                # what the items of the latent factor are
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),                    # what covariates that you want to include in the analysis
                     fitfun = "WLS",
					 exogenous = T)                                                 # and the fit function that you would like to use (WLS is much faster than ML)

```

You can take this object (addFac) which is technically an OpenMx model, and simply fit it using mxRun() 

```ruby
twoFacFit <- mxRun(twoFac)
summary(twoFacFit)
```

This is strongly advised, as it is a great time to test whether the model is being specified the way you want it to be and that you are not getting unrealistic estimates.

Provided that the model looks reasonable, you can plug the model that you have built into the GWAS function using the command below:

```ruby
GWAS(model = twoFac,                                                                 # what model object you would like to fit
	snpData = 'example.pgen',                                                        # that path to the snpData file.
	out="twoFac.log")                                                                # the file that you would like to save the full results into
	                                                                       # the index of the snps (how many) you would like to fit

```

**Note about the snpData argument:** The path to your snpData will likely include switching to a different directory (as you will likely do your analysis in a different folder than your SNP data). All you need to do is point to the data using relative paths. Further, it is able to take plink bed/bim/fam or pgen/psam/pvar data or bgen data (oxford format)

**Note about the SNP argument:** This can be used to run a limited number of SNP (i.e. not the whole snp file). This is particularly useful if you would like to run chop up a chr into several parts without cutting you actual genotype data into separate files.

While GWAS function will take a while and frequently be done on a computing cluster, it is very useful to run a few SNPs (say 10 or 50) in an interactive mode to make sure that: all of your relative file paths to the genotypes are correct, the model is taking a reasonable amount of time (i.e. 1-2 seconds/snp), the SNPs are giving sensible estimates, etc., as the results from a few SNPs can often tell you if there is a problem and that you are running a nonsensical model. 

While most computing clusters have unique queue submission systems (and I would strongly advise you to run GWAS on a computing cluster), the R script that you provide will typically be similar to the one below:

```ruby
require(gwsem)

TwoDat <- read.table("phenoTwoData.txt", header=TRUE)
head(TwoDat)

twoFac <- buildTwoFac(phenoData = TwoDat,                              # data object
                     F1itemNames = c("A1", "A2", "A3"),                # items of the first latent factor
                     F2itemNames = c("B1", "B2", "B3"),                # items of the second latent factor 
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),      # covariates 
                     fitfun = "WLS",
		     exogenous = T)                                                 


GWAS(model = twoFac, snpData = 'example.pgen',  out="twoFac.log") 

```

The next step is to read the results into R. The output from the GWAS function contains all of the estimates and standard errors for all parameters in the model (for each SNP), as well as other general model fitting information.  While you are unlikely to want to read all of this output into R, it is useful to do this on a small test GWAS analysis to ensure that you are getting reasonable results and to familiarize yourself with what is information is available after the analyses, in case you need to explore a specific parameter or model in more detail. This can be done using the command below:  

```ruby
TwoResult <- read.delim("twoFac.log")
```

More likely, you are going to want to read the results into R for specific coefficients that address a particular hypothesis. For the two factor model we might want to look at both regression paths from the SNP to each latent factor as well as the covariance between the latent factors for jointly significant SNPs. To read the data for all three parameters of interest into R we can use the _**loadResults**_ function. This function can be repeated for as many parameters that you want to read into R.  It takes takes two arguments: the path to the data and the column in the results file for the parameter that you want to examine.




```ruby
succinct1 <- loadResults(path = "twoFac.log", focus =  "snp_to_F1")
succinct2 <- loadResults(path = "twoFac.log", focus =  "snp_to_F2")
```

We can then construct a Manhattan plot using the GW-SEM plot() function to visually inspect the general pattern of associations between the SNPs and each factor.

```ruby
plot(succinct1)
plot(succinct2)
```

For the Two Factor Model, we will likely want to examine pleiotropic genetic effects by looking at how the pattern of SNP regression affect the correlation between the latent factors.  We would expect that if a single SNP was associated with both factors, we would see a reduction in the correlation between the factors.  Accordingly, we can sort the SNP results that we read into R by the largest Z statistic to see whether there were any SNPs jointly predicted both latent factors.

To demonstrate the intricacies of the Two Factor GWAS model, when simulating the data, we randomly chose snp141 to be associated with both factors, snp50 to be associated with F<sub>1<\sub> only, and snp719 to be associated with F<sub>2<\sub> only. 

 ```ruby
head(succinct1[order(succinct1$Z, decreasing = T),])
#   MxComputeLoop1 CHR   BP     SNP A1 A2 statusCode catch1  snp_to_F1         Z             P
#1:            142   1  141  snp141  B  A         OK     NA 0.35513118 14.007946  1.393770e-44
#2:             51   1   50   snp50  A  B         OK     NA 0.30405584 11.832143  2.662557e-32
#3:            469   1  468  snp468  B  A         OK     NA 0.08488192  3.406439  6.581637e-04
#4:           1245   1 1244 snp1244  B  A         OK     NA 0.07377348  2.981144  2.871733e-03
#5:            133   1  132  snp132  A  B         OK     NA 0.07040070  2.794059  5.205104e-03
#6:           1718   1 1717 snp1717  A  B         OK     NA 0.06241209  2.507914  1.214464e-02

head(succinct2[order(succinct2$Z, decreasing = T),])
#   MxComputeLoop1 CHR   BP     SNP A1 A2 statusCode catch1  snp_to_F2         Z             P
#1:            142   1  141  snp141  B  A         OK     NA 0.32206276 12.725076  4.290550e-37
#2:            720   1  719  snp719  B  A         OK     NA 0.29285111 11.445648  2.472507e-30
#3:           1836   1 1835 snp1835  B  A         OK     NA 0.06551471  2.616879  8.873773e-03
#4:           1504   1 1503 snp1503  B  A         OK     NA 0.06542646  2.593691  9.495184e-03
#5:           1943   1 1942 snp1942  A  B         OK     NA 0.06326617  2.531957  1.134279e-02
#6:           1047   1 1046 snp1046  B  A         OK     NA 0.06101021  2.412307  1.585193e-02
```

As can be seen, snp141 significantly predicts both latent factors.  Let's take a look at this correlation between the latent factors for this SNP.

First, let's read in the factor correlation results, specifically asking to also load the standard errors, and treating the data object as a data frame (which makes it easier to use base R functions to search through the results). Then lets look at the factor correlation for snp141, as well as snp50 and snp719, and a few other snps (say the first 6 in the results file).

```ruby
facCov <- as.data.frame(loadResults(path = "twoFac.log", focus =  "facCov", .retainSE = T))

facCov[facCov$SNP == "snp141",]

# MxComputeLoop1 CHR  BP    SNP A1 A2 statusCode catch1    facCov   facCovSE           Z            P
#            142   1 141 snp141  B  A         OK     NA 0.3010985 0.01584482    19.00296 1.612025e-80

facCov[facCov$SNP == "snp50",]
# MxComputeLoop1 CHR BP   SNP A1 A2 statusCode catch1    facCov   facCovSE          Z             P
#             51   1 50 snp50  A  B         OK     NA 0.3331034 0.01555011   21.42129 8.462177e-102



facCov[facCov$SNP == "snp719",]
# MxComputeLoop1 CHR  BP    SNP A1 A2 statusCode catch1    facCov   facCovSE          Z             P
#            720   1 719 snp719  B  A         OK     NA 0.3293707 0.01544082   21.33117 5.832476e-101

head(facCov)
#  MxComputeLoop1 CHR BP  SNP A1 A2 statusCode catch1    facCov   facCovSE        Z            P
#1              1   1  0 snp0  B  A         OK     NA 0.3287115 0.01551386 21.18825 1.225600e-99
#2              2   1  1 snp1  A  B         OK     NA 0.3289715 0.01554001 21.16933 1.831431e-99
#3              3   1  2 snp2  B  A         OK     NA 0.3280331 0.01552436 21.13022 4.196174e-99
#4              4   1  3 snp3  A  B         OK     NA 0.3279912 0.01551839 21.13565 3.739797e-99
#5              5   1  4 snp4  A  B         OK     NA 0.3282822 0.01552461 21.14592 3.008364e-99
#6              6   1  5 snp5  B  A         OK     NA 0.3280614 0.01553241 21.12108 5.091485e-99

```

As can be seen, there is a marked reduction on the correlation between the two latent factors for snp141 that is not present for the other SNPs that are associated with only one of the factors, nor for SNPs that are not associated with either factor. Accordingly, we should interpret both the associations with the latent factors and the reduced correlation between the phenotypes when discussing the results.
***

## Suggestions for Conducting the Analysis

Before running any sort of analysis in GW-SEM, users should have a general understanding of the latent traits that they are interested in examining. Specifically, for the one-factor model, the items that load onto the latent variable should be established prior to running the GWAS, or the user risks conducting an analysis that does not test their intended hypothesis.  For example, if the structure of the latent factor differs from the model written in GW-SEM (model misspecification), any conclusions that are drawn from the analysis are questionable.  Alternatively, if there are items that do not load onto the latent factor, this may dilute the GWAS signal.  The best, and most interpretable, results will come from a clearly defined, well specified, phenotypic model. 

***

### Specifying the Model


***
The syntax to fit the two factor model is:

buildTwoFac(phenoData, F1itemNames, F2itemNames, covariates=NULL, ..., fitfun = c("WLS","ML"), minMAF=0.01, modelType=c('RAM','LISREL'))

**phenoData:** the file pathway for the phenotypic data (e.g. "myData.txt" or "phenotype/myData.txt"). This data file can include more variables than those included in the analysis, but GW-SEM will only utilize the items/covariates that are specified. (The dangers of very large dataset is that they can take a long time to load and can take up space in the R environment. This should not affect processing speed for the GWAS analysis, but can create headaches for pre-processing).

**F1itemNames:** a character list of the names of the items that load onto the first latent variable. These names must match variable names in the phenoData file.

**F2itemNames:** a character list of the names of the items that load onto the second latent variable. These names must match variable names in the phenoData file. Note that there can be some overlap between the items that load onto the first and second factor, but to identify the model there must be unique items for each factor.

**covariates:** a character list of covariates that the latent variable will be regressed upon. The default vaule is NULL, but this is really a silly value as typically analysts will include e.g. age, sex, and ancestry principle components in the analysis.

**fitfun:** The fitfun argument specifies which fit function should be used in evaluating the GWAS model. Users may choose between the relatively rapid "WLS", or the slower but asymptotically optimal "ML". In many cases the the differences between the fit functions is trivial and the faster "WLS" option should be used, but in some situations the differences can be quite meaningful (such as when data are Missing at Random - MAR).

**minMAF:** The minimum valid minor allele frequency (MAF). Large differences between the variances of two variables in the same model can cause optimization failures that invalidate the model. Further, very small minor allele frequencies are more affected by outliers or influential observations. Accordingly, users can specify the minimum allowable MAF. The default value is MAF > .01. Users may also wish to filter out small MAF alleles from their genotype files in other software programs, such as PLINK.

**modelType:** GW-SEM models can be specified using either 'RAM' or 'LISREL' algebras. The use of these two interrelated algebraic formulations of SEM specifications provides additional structure to the model specification, reduces unintentional specification errors, and standardizes the model for the GWAS.

##Running the GWAS

The GWAS function takes the model that was created by buildOneFac (or the other prebuilt GW-SEM functions) and conducts a GWAS. To run the GWAS, use the GWAS function. The syntax of the GWAS model is presented below:

GWAS(model, snpData, SNP=NULL, out="out.log")

**model:** an OpenMx model, specified using RAM or LISREL notation. The model argument is designed to take the output from e.g. buildOneFac (or the other prebuilt GW-SEM functions), but advanced users can specify their own arbitrary OpenMx Model.

**snpData:** a pathway to a file containing GWAS data. The data can be in a variety of forms, such as standard PLINK format (bed/bim/fam), PLINK2 format (pgen/pvar/psam), Oxford format (bgen/sample), or CSV format (csv format in much slower due to the lack of compression for non-binary files).

**SNP:** a numerical range that specifies the number of SNPs to be evaluated from the snpData file. This argument can be used to evaluate a subset of snps for model testing. e.g. 1:10 will run the first 10 snps to make sure that the model is functioning the way the users intends, that the files exist pathways are correct. This option is also very useful to specify a range of snps to be evaluated that is smaller than the complete file. For example, users may wish to run several discrete batches of analyses for chromosome 1, by running 1:10000, 100001:200000, etc. This prevents users from constructing numerous snap files for each chromosome. The default value of the SNP argument is NULL, which will run all snps in the file.

**out:** a file name or pathway where the output from the analysis will be saved. The default pathway is "out.log", which will save the file in the working directory. Users should take caution when specifying the output file name so that the output from different analyses/chromosomes do not overwrite existing files.

***


## Fitting the Two Factor Model

As an example, we will fit the two factor model by running the two lines of code presented below.

snpCovs( FacModelData = "Ord1fac.txt", vars = c("y1","y2","y3","y4","x1","x2","x3","x4"), controls = c("c1","c2","c3"), SNPdata = "SNP.txt", output = "Ord2fac") zeroOne = FALSE)
twofacDWLS( itemData = "Ord2fac.txt", snpCov = "Ord2faccov.txt", snpWei = "Ord2facwei.txt", f1Names = c("y1","y2","y3","y4"), f2Names = c("x1","x2","x3","x4"), Controls = c("c1","c2","c3"), output = "Ord2FacOut.txt")


As can be seen, the two factor GW-SEM results present a regression coefficient, standard error, t-statistic and p-value associated with each factor. As with the one factor GW-SEM, it is useful to construct manhattan plots for the p-values for each item. The Manhattan plot for the two factor GW-SEM can be constructed in an analogous way as the residuals GW-SEM above. The code for the two factor GW-SEM Manhattan plot is presented below.

Looking at the Manhattan plots, it appears that there are two hits for each factor. To dig deeper into these associations, we can examine the top set of SNPs for both Factors, as shown below. For simplicity, we will only look at the regression coefficients, the standard errors and the p-values for each factor.

The first set of results sorts the p-values with respect to Factor 1. As we saw on the Manhattan plots there are two genome-wide significant associations for Factor 1. The top hit for Factor 1 is rs2178 and the second hit is rs6001. Looking at the p-values, it is clear that that rs2178 is only associated with Factor 1, as the p-value for the association with Factor 2 is p = .67. The second set of results sorts the p-values with respect to Factor 2. Again there are two significant hits: rs2014 and rs6001. Similar to the Factor 1 case, rs2014 is only associated with Factor 2 (the p-value for its association with Factor 1 is .99). Importantly, rs6001 is associated with both factors. Accordingly, for this example there are SNPs with significant associations: rs2178 is only associated with Factor 1, rs2014 is only associated with Factor 2, and rs6001 is associated with both Factor 1 and Factor 2.

As with the previous models, it is now time to look at the other parameters in the model. For the two factor model, the parameters are stored in "twoFacEstimates.txt". Rather than looking at the first several estimates, this time we will look at the estimates for rs2178, rs2014, and rs6001 (the significant SNPs in the two factor model). These estimates are presented below.

The estimates of the corresponding parameters are very similar for each of the three significant SNPs. As would be expected, the largest difference between these models is that for the rs6001 model, the factor correlation is notably (but subtly) smaller than for the other two models (as shown in the cov12 column of the output), suggesting that the pleiotropic genetic effects are responsible for a part of the association between these latent factors.

Finally, will construct a QQ plot for each factor. The code to conduct these plots is presented below.

For the current example, there is no serious inflation of the test statistics relative to what would be expected under the null hypothesis. Note the two points that are substantially above the line in the top right corners of both QQ plots. These denote the significant SNPs for both factors.

A script to run all of the analyses for the residuals model can be found here. 

***


