# The Residuals Model

The Residuals GWAS Model is a variant of the [One Factor GWAS Model](https://github.com/bradverhulst/gwsem/wiki/One-Factor-Model) where the SNP directly predicts one or more of the item-specific residuals, after accounting for the latent variable.  The Figure above presents a schematic representation of the Residuals Model. Importantly, the same identification requirements are the same as the one factor GW-SEM model above. The notable difference is that the individual items are regressed on each SNP (&gamma;<sub>k</sub>). The residuals model allows researchers to regress a SNP on the residuals of the items that contribute to a latent factor. It is also possible to fit the residuals model so that the latent factor and a subset of the items are regressed on the SNPs. If you do this, in order to identify the SEM it is only possible to regress a subset of the items and the latent factor on the SNP or the model will not be identified. 

In the Figure below , the latent factor (F<sub>1</sub>) is the latent factor, and the observed items (x<sub>k</sub>) are indicators of the latent factor. The association between the latent factor and the observed indicators are estimated by the factor loadings (λ<sub>k</sub>). The residual variances (δ<sub>k</sub>) indicate the remaining variance in (x<sub>k</sub>) that is not shared with the latent factor. The regression of the observed items on the SNP (for all SNPs in the analysis) is depicted by (γ<sub>k</sub>). 

![A schematic depiction of the Residuals Model](https://github.com/bradverhulst/gwsem/blob/master/exampleFigs/resid.jpg)

## Example

To demonstrate how to use the Residuals model function, we use the same simulated data that we used for the one factor GWAS model example. Specifically, we simulated data for 6,000 individuals consisting of 2,000 SNPs, three correlated items and 6 covariates (that are a proxy for age, sex, ancestry principle components or other confounds). The simulated data are loosely based upon the data used in the demonstration of Pritikin et al (under review). In the paper, we examined the association a latent substance use frequency variable that used tobacco, cannabis, and alcohol as indicators. The latent variable was regressed on each SNP with a minor allele frequency larger than 1%. The phenotypic data data and scripts that we used in this demonstration can be found [here](https://github.com/bradverhulst/gwsem/tree/master/gwsemOneFacExample). The GWAS data can be found [here](https://github.com/bradverhulst/gwsem/tree/master//Example%20GWAS%20Data).


## Running the One-Factor Residuals GWAS

**Note: Many of these steps are similar to those that you would take to run the One Factor GWAS.**

The first step is opening R and loading the GW-SEM package into the R computing environment (which will also load all of the dependencies, such as OpenMx), which can be achieved by using the command below:

```ruby
require(gwsem)
```

The next step is to load the phenotypic data. This will include any items that define the latent variable or necessary covariates that you will include in the analysis, such as age, sex, or ancestry principle components. Once the phenotypic data has been loaded, it is advisable to make sure the data has been loaded properly. This can be done in a number of ways. For example, you can use the head command to inspect the first few lines of data, look at the means and variances, construct tables of ordinal variables, etc.  If your data didn't load properly, it will likely cause problems for the analysis and for the interpretation of the results.

```ruby
phenoData <- read.table("phenoData.txt", header=TRUE)
head(phenoData)
```

Once the data is loaded into R, you can recode the data, transform it, and tell R if we have ordinal or binary indicators using mxFactor(). The data for this example are simulated and continuous, and therefore, we will not be doing anything now, but if you would like to chop the indicators up into binary or ordinal variables, this would be when you would do it.


After you are satisfied the data is in the appropriate shape we can build a one factor GWAS model with the following command: 

```ruby
                                                                                       # You must tell GW-SEM:
addFacRes <- buildOneFacRes(phenoData,                                                 # what the data object is (which you read in above)
                     itemNames = c("tobacco", "cannabis", "alcohol"),                  # what the items of the latent factor are
                     res = c("tobacco", "cannabis", "alcohol"),                        # the items you want to regress onto the SNP (the default is all of the items)
                     factor = F,                                                       # factor is set to F by default
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),                      # what covariates that you want to include in the analysis
                     fitfun = "WLS")                                                   # and the fit function that you would like to use (WLS is much faster than ML)

```

**Note about regressing the latent factor and SOME of the items on the SNPs:** The model is not identified if both the latent variable and all of the items are simultaneously regressed on the SNP. In some cases, however, it may be important to understand the effect of the SNP on both the latent factor and the items. In such cases, you must remove at least path from the SNP to the item residuals, and then it is possible to toggle the _factor_ argument from FALSE to TRUE.

You can take this object (addFacRes) which is technically an OpenMx model, and simply fit it using mxRun() 

```ruby
addFacResFit <- mxRun(addFacRes)
summary(addFacFit)
```
This is strongly advised, as it is a great time to test whether the model is being specified the way you want it to be and that you are not getting unrealistic estimates.

Provided that the model looks reasonable, you can plug the model that you have built into the GWAS function using the command below:

```ruby
GWAS(model = addFacRes,                                                                 # what model object you would like to fit
	snpData = 'example.pgen',                                                        # that path to the snpData file.
	out="FacRes.log",                                                                # the file that you would like to save the full results into
	SNP=1:200)                                                                       # the index of the snps (how many) you would like to fit
```

While GWAS function will take a while and frequently be done on a computing cluster, it is very useful to run a few SNPs (say 10 or 50) in an interactive mode to make sure that: all of your relative file paths to the genotypes are correct, the model is taking a reasonable amount of time (i.e. 1-2 seconds/snp), the SNPs are giving sensible estimates, etc., as the results from a few SNPs can often tell you if there is a problem and that you are running a nonsensical model. 

While most computing clusters have unique queue submission systems (and I would strongly advise you to run GWAS on a computing cluster), the R script that you provide will typically be similar to the one below:

```ruby
require(gwsem)

phenoData <- read.table("phenoData.txt", header=TRUE)
head(phenoData)

addFacRes <- buildOneFacRes(phenoData,                                                 # what the data object is (which you read in above)
                     c("tobacco", "cannabis", "alcohol"),                              # what the items of the latent factor are
                     covariates=c('pc1','pc2','pc3','pc4','pc5'),                      # what covariates that you want to include in the analysis
                     fitfun = "WLS")                                                   # and the fit function that you would like to use (WLS is much faster than ML)

GWAS(model = addFacRes,                                                                 # what model object you would like to fit
	snpData = 'example.pgen',                                                        # that path to the snpData file.
	out="latFac.log")                                                                # the file that you would like to save the full results into
```

The next step is to read the results into R. The output from the GWAS function contains all of the estimates and standard errors for all parameters in the model (for each SNP), as well as other general model fitting information.  While you are unlikely to want to read all of this output into R, it is useful to do this on a small test GWAS analysis to ensure that you are getting reasonable results and to familiarize yourself with what is information is available after the analyses, in case you need to explore a specific parameter or model in more detail. This can be done using the command below:  

```ruby
FullResult <- read.delim("latFac.log")
```

More likely, you are going to want to read the results into R for SNP regression coefficients. This function takes takes two arguments: the path to the data and the column in the results file for the parameter that you want to examine.

Now that we have run the residual's model, we have multiple parameters that we want to load into R (Three in this case). The function to do this is the same as for one parameter, but you need to do it three times (once for each parameter of interest). This gives several R objects.

```ruby
succinctTob <- loadResults(path = "facRes.log", focus =  "snp_to_tobacco")
succinctCan <- loadResults(path = "facRes.log", focus =  "snp_to_cannabis")
succinctAlc <- loadResults(path = "facRes.log", focus =  "snp_to_alcohol")
```

# Now we can plot all the residual manhattan plots

We can then construct a Manhattan plot using the GW-SEM plot() function. 

```ruby
plot(succinctTob)
plot(succinctCan)
plot(succinctAlc)
```

The loadResults() function formats the summary statistics from the analyses so that the data object can be seamlessly used by qqman, LDhub, and other post GWAS processing software.

***

### Using the Functions and Understanding the Arguments

The GW-SEM workflow is split up into two primary functions. The first function specifies the model and the second function runs the GWAS analysis.

### buildOneFacRes 

```ruby
buildOneFacRes(phenoData, itemNames, factor = F, res = itemNames, covariates=NULL, ..., fitfun = c("WLS","ML"), minMAF=0.01, gxe = NULL, exogenous = NA)
```

**phenoData:** the file pathway for the phenotypic data (e.g. "myData.txt" or "phenotype/myData.txt"). This data file can include more variables than those included in the analysis, but GW-SEM will only utilize the items/covariates that are specified. (The dangers of very large dataset is that they can take a long time to load and can take up space in the R environment. This should not affect processing speed for the GWAS analysis, but can create headaches for pre-processing).

**itemNames:** a character list of the names of the items that load onto the latent variable. These names must match variable names in the phenoData file.

**factor:** is a logical statement indicating whether or not the SEM will include a regression of the latent factor on the snp. The default is to not regress the latent variable on the snp.

**res:** is a character list of the names of the items where the residual will be regress onto the snp. The default is to regress all of the items onto the snp, but in doing so it is not possible to regress the latent factor onto the snp.

**covariates:** a character list of covariates that the latent variable will be regressed upon. The default vaule is NULL, but this is really a silly value as typically analysts will include e.g. age, sex, and ancestry principle components in the analysis.

**fitfun:** The fitfun argument specifies which fit function should be used in evaluating the GWAS model. Users may choose between the relatively rapid "WLS", or the slower but asymptotically optimal "ML". In many cases the the differences between the fit functions is trivial and the faster "WLS" option should be used, but in some situations the differences can be quite meaningful (such as when data are Missing at Random - MAR).

**minMAF:** The minimum valid minor allele frequency (MAF). Large differences between the variances of two variables in the same model can cause optimization failures that invalidate the model. Further, very small minor allele frequencies are more affected by outliers or influential observations. Accordingly, users can specify the minimum allowable MAF. The default value is MAF > .01. Users may also wish to filter out small MAF alleles from their genotype files in other software programs, such as PLINK.

**gxe:** The observed variable name that will be used to moderate (interact with) the effect of the SNP on the phenotypes.  For example, you may want to moderate the SNP regression by sex.  In this situation, you would specify **_gxe = "sex"_** and add _**"snp_sex"**_ to the list of covariates.

**exogenous:** This argument specifies how you would like to integrate the covariates into the analysis. If **exogenous = T**, each items will be directly regressed on each covariate. If **exogenous = F**, the latent factor(s) will be directly regressed on each covariate. Setting **exogenous = T** does not assume that the items are related to the covariates proportional to their factor loadings (which is probably preferable in most cases).

## Running the GWAS

The GWAS function takes the model that was created by **buildOneFacRes** (or the other prebuilt GW-SEM functions) and conducts a GWAS. To run the GWAS, use the GWAS function. The syntax of the GWAS model is presented below:

```ruby
GWAS(model, snpData, out="out.log", SNP=NULL, startFrom = 1L)
```

**model:** an OpenMx model, specified using RAM or LISREL notation. The model argument is designed to take the output from e.g. buildOneFac (or the other prebuilt GW-SEM functions), but advanced users can specify their own arbitrary OpenMx Model.

**snpData:** a pathway to a file containing GWAS data. The data can be in a variety of forms, such as standard PLINK format (bed/bim/fam), PLINK2 format (pgen/pvar/psam), Oxford format (bgen/sample), or CSV format (csv format in much slower due to the lack of compression for non-binary files).

**SNP:** a numerical range that specifies the number of SNPs to be evaluated from the snpData file. This argument can be used to evaluate a subset of snps for model testing. e.g. 1:10 will run the first 10 snps to make sure that the model is functioning the way the users intends, that the files exist pathways are correct. This option is also very useful to specify a range of snps to be evaluated that is smaller than the complete file. For example, users may wish to run several discrete batches of analyses for chromosome 1, by running 1:10000, 100001:200000, etc. This prevents users from constructing numerous snap files for each chromosome. The default value of the SNP argument is NULL, which will run all snps in the file.

**out:** a file name or pathway where the output from the analysis will be saved. The default pathway is "out.log", which will save the file in the working directory. Users should take caution when specifying the output file name so that the output from different analyses/chromosomes do not overwrite existing files.



*** 
